src/calculate_bucket.hpp

```diff
@@ -154,69 +159,27 @@ class F1Calculator {
         return std::make_pair(CalculateF(L), L);
     }
 
-    // Returns an evaluation of F1(L), and the metadata (L) that must be stored to evaluate F2,
-    // for 'number_of_evaluations' adjacent inputs.
-    inline std::vector<std::pair<Bits, Bits>> CalculateBuckets(
-        const Bits& start_L,
-        uint64_t number_of_evaluations)
+    // F1(x) values for x in range [first_x, first_x + n) are placed in res[].
+    // n must not be more than 1 << kBatchSizes.
+    void CalculateBuckets(uint64_t first_x, uint64_t n, uint64_t *res)
     {
-        uint16_t num_output_bits = k_;
-        uint16_t block_size_bits = kF1BlockSizeBits;
-
-        uint64_t two_to_the_k = (uint64_t)1 << k_;
-        if (start_L.GetValue() + number_of_evaluations > two_to_the_k) {
-            throw "Evaluation out of range";
-        }
-        // Counter for the first input
-        uint64_t counter = (start_L.GetValue() * (uint128_t)num_output_bits) / block_size_bits;
-        // Counter for the last input
-        uint64_t counter_end =
-            ((start_L.GetValue() + (uint128_t)number_of_evaluations + 1) * num_output_bits) /
-            block_size_bits;
-
-        std::vector<Bits> blocks;
-        uint64_t L = (counter * block_size_bits) / num_output_bits;
-        uint8_t ciphertext_bytes[kF1BlockSizeBits / 8];
-
-        // Evaluates ChaCha8 for each block
-        while (counter <= counter_end) {
-            chacha8_get_keystream(&this->enc_ctx_, counter, 1, ciphertext_bytes);
-            Bits ciphertext(ciphertext_bytes, block_size_bits / 8, block_size_bits);
-            blocks.push_back(std::move(ciphertext));
-            ++counter;
+        uint64_t start = first_x * k_ / kF1BlockSizeBits;
+        // 'end' is one past the last keystream block number to be generated
+        // We first find the last used bit of the keystream, then convert it to
+        // block number and add one
+        uint64_t end = ((first_x + n) * k_ - 1) / kF1BlockSizeBits + 1;
+        uint64_t n_blks = end - start;
```

Comment:
I think it would make sense to have:

```
assert(n_blks <= cdiv((k_ + kExtraBits) << kBatchSizes, kF1BlockSizeBits) + 2);
```

Reply:
I did the math again and realized that this computation is wrong. I'll submit a new version with an assert here (a different one).

Reply:
Please see the new version with the assert. I'm reasonably sure that the math is right now and it's sufficient to ensure that `n <= (1U << kBatchSizes)`.